File: packages/js/sdk/src/wrap-tool-handlers.ts
/**
 * wrapToolHandlers — creates a TrikAgent for tool-mode triks.
 *
 * Tool-mode triks export native tools to the main agent (no handoff, no session).
 * Each handler receives the validated input and returns structured output.
 */

import type { TrikAgent, TrikContext, ToolExecutionResult } from '@trikhub/manifest';

/**
 * A single tool handler function.
 * Receives validated input and trik context, returns structured output.
 */
export type ToolHandler = (
  input: Record<string, unknown>,
  context: TrikContext,
) => Promise<Record<string, unknown>> | Record<string, unknown>;

/**
 * Wrap a map of tool handlers into a TrikAgent.
 *
 * @example
 * ```typescript
 * import { wrapToolHandlers } from '@trikhub/sdk';
 *
 * export default wrapToolHandlers({
 *   getWeather: async (input, context) => {
 *     const { city } = input as { city: string };
 *     return { temperature: 22, unit: 'celsius', condition: 'sunny' };
 *   },
 * });
 * ```
 */
export function wrapToolHandlers(
  handlers: Record<string, ToolHandler>,
): TrikAgent {
  return {
    async executeTool(
      toolName: string,
      input: Record<string, unknown>,
      context: TrikContext,
    ): Promise<ToolExecutionResult> {
      const handler = handlers[toolName];
      if (!handler) {
        throw new Error(
          `Unknown tool "${toolName}". Available tools: ${Object.keys(handlers).join(', ')}`,
        );
      }

      const output = await handler(input, context);
      return { output };
    },
  };
}

File: packages/python/trikhub/sdk/wrap_tool_handlers.py
"""
Wrap tool handler functions into a TrikAgent with executeTool support.

Simple dispatch: looks up the handler by tool name, calls it, wraps the result.
Mirrors packages/js/sdk/src/wrap-tool-handlers.ts.
"""

from __future__ import annotations

import asyncio
from typing import Any, Callable, Union

from trikhub.manifest import TrikContext, ToolExecutionResult

ToolHandler = Callable[
    [dict[str, Any], TrikContext],
    Union[dict[str, Any], Any],  # sync or async
]


class _WrappedToolHandlers:
    """TrikAgent implementation wrapping tool handler functions."""

    def __init__(self, handlers: dict[str, ToolHandler]) -> None:
        self._handlers = handlers

    async def execute_tool(
        self,
        tool_name: str,
        input: dict[str, Any],
        context: TrikContext,
    ) -> ToolExecutionResult:
        """Execute a tool by name.

        Args:
            tool_name: Name of the tool to execute
            input: Tool input parameters
            context: Trik execution context

        Returns:
            ToolExecutionResult with the tool's output

        Raises:
            ValueError: If tool_name is not in the handlers dict
        """
        handler = self._handlers.get(tool_name)
        if handler is None:
            available = ", ".join(sorted(self._handlers.keys()))
            raise ValueError(
                f"Unknown tool '{tool_name}'. Available tools: {available}"
            )

        result = handler(input, context)
        # Support async handlers
        if asyncio.iscoroutine(result):
            result = await result

        return ToolExecutionResult(output=result)


def wrap_tool_handlers(handlers: dict[str, ToolHandler]) -> _WrappedToolHandlers:
    """Wrap a dict of tool handler functions into a TrikAgent.

    Each handler receives (input, context) and returns a dict output.
    Handlers can be sync or async.

    The returned object implements the TrikAgent protocol with `execute_tool()`.

    Args:
        handlers: Dict mapping tool names to handler functions

    Returns:
        A TrikAgent-compatible object with execute_tool()
    """
    return _WrappedToolHandlers(handlers)
